src/itwinai/loggers.py

```diff
@@ -604,7 +604,7 @@ def log(
             if not isinstance(item, str):
                 # Save the object locally and then log it
                 name = os.path.basename(identifier)
```

Comment:
Regarding this whole function: 

It seems like we do a bunch of pattern matching on kind and item. This might be streamlined by using 

```python
match (kind, item):
            case ("metric", _):
                mlflow.log_metric(key=identifier, value=item, step=step)

            case ("artifact", str()):
                mlflow.log_artifact(local_path=item, artifact_path=identifier)

            case ("artifact", _):
                name = os.path.basename(identifier)
                save_path = Path(self.savedir) / ".trash" / str(name)
                save_path.mkdir(os.path.dirname(save_path), exist_ok=True)
                item = self.serialize(item, save_path)
                mlflow.log_artifact(local_path=item, artifact_path=identifier)
                ...
```
Not super important, but we lose a bit of indentation in which case, making the code a bit more readable.

Reply:
cool, I missed this feature!

Reply:
tried it, didn't like it for complex conditions, e.g., `not isinstance(item, str)`

Reply:
I haven't used pattern matching too much in Python, so for the exercise I wrote out what I think the code would look like. I think you can stay at one indentation level if you do this, though I haven't actually tested the code: 

```python
match (kind, item): 
            case ("metric", _): 
                mlflow.log_metric(key=identifier, value=item, step=step)
            case ("artifact", str()): 
                mlflow.log_artifact(local_path=item, artifact_path=identifier)
            case ("artifact", _): 
                name = os.path.basename(identifier)
                save_path = Path(self.savedir) / ".trash" / name
                save_path.mkdir(os.path.dirname(save_path), exist_ok=True)
                item = self.serialize(item, save_path)
                mlflow.log_artifact(local_path=item, artifact_path=identifier)
            case ("model", item) if isinstance(item, torch.nn.Module): 
                import torch
                mlflow.pytorch.log_model(item, identifier)
            case ("model", _): 
                import torch
                print("WARNING: unrecognized model type")
            case ("dataset", item) if isinstance(item, mlflow.data.Dataset): 
                mlflow.log_input(item)
            case ("dataset", _): 
                print("WARNING: unrecognized dataset type. " "Must be an MLFlow dataset")
            case ("torch", _): 
                import torch

                # Save the object locally and then log it
                name = os.path.basename(identifier)
                save_path = Path(self.savedir) / ".trash" / name
                save_path.mkdir(os.path.dirname(save_path), exist_ok=True)
                torch.save(item, save_path)
                # Log into mlflow
                mlflow.log_artifact(local_path=save_path, artifact_path=identifier)
            case ("dict", _): 
                mlflow.log_dict(dictionary=item, artifact_file=identifier)
            case ("figure", _): 
                mlflow.log_figure(
                    artifact_file=identifier,
                    figure=item,
                    save_kwargs=kwargs.get("save_kwargs"),
                )
            case ("image", _): 
                mlflow.log_image(artifact_file=identifier, image=item)
            case ("param", _): 
                mlflow.log_param(key=identifier, value=item)
            case ("text", _): 
                mlflow.log_text(artifact_file=identifier, text=item)
            case _: 
                pass 

```